pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -225,27 +230,75 @@ func configFromEnv() (cfg Config, ok bool) {
 		cfg.Global.DomainName = os.Getenv("OS_USER_DOMAIN_NAME")
 	}
 
-	ok = cfg.Global.AuthURL != "" &&
-		cfg.Global.Username != "" &&
-		cfg.Global.Password != "" &&
-		(cfg.Global.TenantID != "" || cfg.Global.TenantName != "" ||
-			cfg.Global.DomainID != "" || cfg.Global.DomainName != "" ||
-			cfg.Global.Region != "" || cfg.Global.UserID != "" ||
-			cfg.Global.TrustID != "")
-
 	cfg.Metadata.SearchOrder = fmt.Sprintf("%s,%s", configDriveID, metadataID)
 	cfg.Metadata.DHCPDomain = "novalocal"
 	cfg.BlockStorage.BSVersion = "auto"
+}
 
-	return
+func hasKey(m map[string][]byte, key string) bool {
+	if _, ok := m[key]; ok {
+		return true
+	}
+	return false
 }
 
-func readConfig(config io.Reader) (Config, error) {
-	if config == nil {
-		return Config{}, fmt.Errorf("no OpenStack cloud provider config file given")
+func configFromSecret(cfg *Config, kubeClient clientset.Interface, name string, namespace string) error {
+	secret, err := kubeClient.CoreV1().Secrets(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+	} else {
+		// When a secret is specified with some of the keys below, we should make sure it has a specific
+		// user id/name and should not use the global admin credentials.
+		specialKeys := []string{"tenant-id", "tenant-name", "region"}
```

Comment:
sorry to keep pushing on this, but this overlaying still seems problematic. why wouldn't we require them to provide credentials if they want to point to different things than the cluster-admin configured? some currently allowed examples that seem suspicious:
* they can swap a different user-id/username and keep the cluster-admin specified password
* they can swap a different domain-id/domain-name and keep the cluster-admin specified username/password